### ISSUE TITLE
Get final uncompressed size

### ISSUE THREAD

**Issue by jviotti** [addaleax/lzma-native#15]
Hi, is there a way to tell the final uncompressed size of a `*.xz` before attempting to do so with this module?

**Comment by addaleax** [comment]
Sorry, I don’t think so… I’ll look into it when I find the time. Just out of curiosity, what is your use case for that?

**Comment by jviotti** [comment]
@addaleax I'm working at [Etcher](http://www.etcher.io), an image writer built with web technologies. We want to support writing `xz` compressed images by uncompressing the stream and piping to the device, however in order to get a nice progress bar, we need to know the final uncompressed size.

I'm looking into `liblzma` myself as well, so I might be able to contribute something.

The uncompressed size is shown when running `xz --list <file>`. By taking a closer look, this is calculated by running the `lzma_index_uncompressed_size()` function on a `lzma_index` populated by a quite big `parse_indexes()` function from the `xz` command line tool.

**Comment by addaleax** [comment]
Oh, yeah, there’s [this](https://github.com/addaleax/xz/blob/master/src/xz/list.c#L146-L148) beautiful `TODO:` comment… :smile: If you can contribute anything here, or possibly even upstream at `liblzma`, that would be so awesome!

In any case, I know that other users of this library have decided to compute progress bars for decompressing depending on the compressed file size, not the uncompressed size. That’s clearly not as beautiful, but I guess it’s okay as an indicator.

**Comment by jviotti** [comment]
> In any case, I know that other users of this library have decided to compute progress bars for decompressing depending on the compressed file size, not the uncompressed size. That’s clearly not as beautiful, but I guess it’s okay as an indicator.

Sadly there is another edge case I didn't consider: After we perform the write, we read-back the same amount of bytes we wrote from the device in order to calculate CRC32 checksums for write validation reasons, so I'd still need to real size somehow. I could accumulate what was written to the device to get that number, but it'd be very cool to just have the correct size before hand.

***

I'll take a closer look at `lzma` and see what could be done. I honestly have little knowledge on how the lzma compression works, but I guess that can be fixed :)

In the worst case we could re-use part of that function to parse indexes in this module.

**Comment by addaleax** [comment]
Just so you know, the file size as indicated by the index can be messed with. I don’t know if any encoders do that or what the implications of that would be, but it’s not completely reliable when the input may come from untrusted sources. (Don’t know whether that applies to you).

If you need any help with the coding stuff, let me know – if you manage to separate the I/O stuff in `parse_indexes` from the calls to `lzma_*` then the largest buch of work if probably already done.

**Comment by jviotti** [comment]
> Just so you know, the file size as indicated by the index can be messed with. I don’t know if any encoders do that or what the implications of that would be, but it’s not completely reliable when the input may come from untrusted sources. (Don’t know whether that applies to you).

I see. According to the specification, the uncompressed size in bytes is also stored in each block. An alternative implementation would be to iterate through every block to accumulate the final size, but sounds very wasteful, and could be slow on very large files.

> If you need any help with the coding stuff, let me know – if you manage to separate the I/O stuff in parse_indexes from the calls to lzma_* then the largest buch of work if probably already done.

Cool, thanks, I will. For now, I'm decoupling `parse_indexes` from `xz_file_info` (so we can operate on `lzma_index` directly) and cutting down a lot of things that we don't need for this use case (like the compression ratio, etc).

**Comment by addaleax** [comment]
> I see. According to the specification, the uncompressed size in bytes is also stored in each block. An alternative implementation would be to iterate through every block to accumulate the final size, but sounds very wasteful, and could be slow on very large files.

Agreed, that sounds like a lot of work (both in terms of coding and processing). Plus, I’ve made the experience that it’s quite confusing for everyone when different tools give different answers for the same file, so I’d prefer to stick with what `xz -l` says.

**Comment by addaleax** [comment]
Oh, and btw, there are (yet undocumented, but stable) `stream.totalIn()` and `stream.totalOut()` methods for the streams created by this module. They require actually decompressing the file, of course, but you can check against those values if you’re doing that anyway.

**Comment by jviotti** [comment]
Thanks @addaleax ,

I'm going to follow your suggestion on making use of the compressed size for the progress bar, and using `stream.totalOut()` for the validation step since I'll need to get the xz compression working with an archive streamed over HTTP. In any case, I'd love to contribute back to this project for the sake of improving my C skills, and for fun of course :)

After studying `parse_indexes()` in more depth, I have the following questions:

- You mentioned the need to decouple this function from I/O. 

As far as I can see `liblzma` doesn't bother with I/O, this being the responsibility of the client, like `xz` for example. `parse_headers()` actually seems to parse the whole file backwards, and uses I/O to, for example, parse the footer to determine the actual location of the index, and keeps going to decode the stream, etc. 

I can't think of a way to decouple this from I/O, since the file as a whole is needed to get the information we want. Did I misinterpret what you meant?

- The `parse_indexes()` function supports decoding multiple indexes. This sounded weird at first given that `xz` only compresses a single file, but I saw the following on the specification:

> While a typical file contains only one Stream and no Stream
> Padding, a decoder handling standalone .xz files SHOULD support
> files that have more than one Stream or Stream Padding.

This this something his module supports? Can you show me an example of how to create such a file?

- The TODO comment you pointed to says:

> liblzma should have a function that takes a callback function to parse the Index(es) from a .xz file to make it easy for applications.

Being a JavaScript developer, I might be biased regarding what the meaning of callback is, but as far as I understand, the comment means that `liblzma` should have a function that reads the `xz` file and allows the client to pass a pointer to a function that gets passed the file index to perform the parsing. The callback is called as many times as an index is found, and all of them are concatenated into a final `lzma_index` struct.

However this conflicts with the fact that `liblzma` doesn't seem to have I/O as a responsibility.

***

In conclusion, the responsibilities of `parse_headers()` fall in the following categories:

- Find the position of the index.
- Parse the index.
- Parse the remaining of the file in order to see if there is another index (e.g: a file with multiple streams).
- Concatenate all found indexes (with `lzma_index_cat()`).

All these items require I/O (except the last one), so I can't think of how we can move any functionality to `liblzma`. Am I missing something?

Thanks for the patience and sorry if any confusion is due to my lack of extensive experience in C :)

**Comment by addaleax** [comment]
> As far as I can see liblzma doesn't bother with I/O, this being the responsibility of the client, like xz for example.

That’s right – I was mainly talking about the `io_pread` calls in `parse_indexes()` (in `list.c`).

> This this something his module supports? Can you show me an example of how to create such a file?

Yes, but thanks for reminding me that there can be stream padding – I can’t *believe* I forgot that that when writing the code for it in this module. It’s fixed now, though. :smile:
If you’re a bit unix-y, you can do something like this:
```bash
$ echo ABC | xz > multi-part.xz
$ echo DEF | xz >> multi-part.xz
$ echo GHI | xz >> multi-part.xz
$ unxz < multi-part.xz 
ABC
DEF
GHI
```

And otherwise, there’s also the section in this library’s tests that shows how to do this a bit more programatically: https://github.com/addaleax/lzma-native/blob/9bba75e8a72b5408398871846b2f21ae847042d1/test/stream.js#L517

> The TODO comment you pointed to says:
> > liblzma should have a function that takes a callback function to parse the Index(es) from a .xz file to make it easy for applications.

Callback means pretty much the same as it does in JS. What I think it means is that instead of calling `io_pread` directly, `parse_indexes` should be rewritten to take an callback that does the I/O for it. In Node.js, that might be something like `fs.read`, but it might be nice to have that overridable by the user.

There’s definitely a challenge here in writing this in a way such that it plays nicely with async I/O that I haven’t fully thought of so far. It’s probably easiest to keep it synchronous for now and then move on from that point…

In my head, a sync version might look something like this:

In liblzma (*):
```c
// Function that reads data for consumption by liblzma.
typedef void (*read_function)(void* opaque, void* buffer, size_t count, off_t offset);

void lzma_parse_indexes(xz_file_info* xfi, read_function read_fn, void* opaque, size_t size) {
    // essentially the same as parse_indexes from list.c
    // with `size` used instead of `pair->src_st.st_size`
    // and read_fn(opaque, &buf, …, pos) instead of io_pread(pair, &buf, …, pos)
}
```

Then one could use that function to parse the indexes while keeping the I/O stuff to oneself, e.g. a Unix user might use a wrapped version of `pread(2)` for `read_fn`.

Re-writing this for async stuff would be the next step, but with systems like Node.js that’s (luckily?) unavoidable. It would mean that memory management and such come into place, and I don’t know yet how that would be handled best (it’s definitely possible, though).

* Or in lzma-native, at least for now. But if we get this done right and you are fine with publishing eventual code as public domain, I’d love to get this pitched to Lasse Collin for inclusion in the actual liblzma distribution.

> Thanks for the patience and sorry if any confusion is due to my lack of extensive experience in C :)

Glad to help! :)

**Comment by addaleax** [comment]
@jviotti Have you had the chance to look into this further? Otherwise I can take some time to spend on this now :)

**Comment by jviotti** [comment]
Hi @addaleax ,

Yeah, been doing a bit of work on Sunday, I'll have something to push today!

**Comment by jviotti** [comment]
@addaleax I pushed a WIP branch to https://github.com/jviotti/xz where the decoupling is taking place.

See the following commit: https://github.com/jviotti/xz/commit/71580429d609013cdc4209c5782a085979fc389e

`parse_indexes()` is almost decouples completely from I/O, with the exception of `io_buf`, where I don't have a clear idea on how to keep moving forward.

`io_buf` is defined as follows:

```c
typedef union {
	uint8_t u8[IO_BUFFER_SIZE];
	uint32_t u32[IO_BUFFER_SIZE / sizeof(uint32_t)];
	uint64_t u64[IO_BUFFER_SIZE / sizeof(uint64_t)];
} io_buf;
```

The `io_pread()` function seems to store data as an `uint32_t` buffer, but the `liblzma` uses a `uint8_t` buffer. Not sure how to encapsulate these differences in a way such that `parse_indexes()` doesn't care about the internal structure of the buffer it writes to. Ideas?

Also, `parse_indexes()` takes a `xz_file_info` struct, which xz specific (not liblzma). I'd be nice if it'd take a `lzma_index`. The only extra field assigned on `xz_file_info` is the stream padding. I guess we can introduce another parameter to `parse_indexes()` to take a pointer to a `uint64_t` for the padding, but I'm sure there is a better way.

**Comment by addaleax** [comment]
I think it’s safe to drop `io_buf buf;` inside the method and replace it with `uint8_t buf[8192]` (and use `uint8_t*` as the parameter type for `parse_indexes_read_callback`). There is only one place where the `u32` field is actually used in the code here, and a simple change from `if (buf.u32[i] != 0)` to `if (((uint32_t*)buf)[i] != 0)` should take care of that.

Regarding `xz_file_info` – yeah, I guess that would have to be separated out, too. My first instinct is dissolving that into a `lzma_index**` and a `uint64_t*`, too, but I’m starting to feel that it may be more appropriate to group all these parameters into a new `struct` type.

That might make it easier to switch to async code later, too… and I’m not sure what to do about the error handling stuff yet, but I think it would help.

**Comment by jviotti** [comment]
@addaleax 

Awesome, thanks for the feedback. I updated the commit referenced with both changes. Let me know if you have any comments on the changes so far, so I update the branch from within `xz`, which is easier to test than directly in `lzma-native`.

Once everything is in place, the next step would be to move this functionality to `lzma-native` (unless you can to get this into `liblzma` and contribute upstream directly. I'm fine with both, since I don't need this functionality right now). 

In case we want to integrate to `lzma-native` first, I was thinking on having a new top level function called `lzma.getUncompressedSize()` which takes a file descriptor (which is more flexible than a filename) and a function to perform the write (synchronous for now).

Regarding asynchronicity, when would you like to start work on this? As far as I understand, `libuv` would provide this functionality, but we would need to do it from within this module, and not `liblzma`. So `liblzma` would contain a synchronous version and `lzma-native` simply wraps the sync function with `libuv`.

**Comment by addaleax** [comment]
>  so I update the branch from within xz, which is easier to test than directly in `lzma-native`.

That is definitely the easiest thing to do. :+1:

> Once everything is in place, the next step would be to move this functionality to lzma-native (unless you can to get this into liblzma and contribute upstream directly. I'm fine with both, since I don't need this functionality right now).

:+1: Lasse Collin does a fantastic job of mainting the upstream library, and I’ll definitely suggest that, but there’s no need to wait for a new `liblzma` release.

> In case we want to integrate to lzma-native first, I was thinking on having a new top level function called lzma.getUncompressedSize() which takes a file descriptor (which is more flexible than a filename) and a function to perform the write (synchronous for now).

Heh, that’s an easy way to take care of any sync/async trouble. :smile: But why not, it’ll definitely get you the feature faster.

> Regarding asynchronicity, when would you like to start work on this?

As soon as you say you’re done with your part, and that’s completely up to you. :)

> As far as I understand, libuv would provide this functionality, but we would need to do it from within this module, and not liblzma. So liblzma would contain a synchronous version and lzma-native simply wraps the sync function with libuv.

Nah, libuv doesn’t have to play a role in here (at least not explicitly). At least right now in my head, it would look like this:
```
[JS] lzma.getUncompressedSize(read_callback)
… calls [lzma-native] some native function wrapper for getUncompressedSize
… calls [liblzma] parse_indexes(…)
… calls [lzma-native] a wrapper for the JS callback
… calls [JS] The original read_callback, e.g. something based on fs.read(fd, …)
… ??? somehow tell liblzma about the new data
```

The tricky question here is how to get the read data from the callback to the original `parse_indexes` function. That may be quite a bit of work, because it might ultimately mean storing **all** state from `parse_indexes` in some `struct` and having a way to resume the work at some point inside the function.

You don’t have to worry about that unless you’re eager to do a lot more C coding, I’ll be glad to it. :smile:

As far as your code changes are concerned: The name `lzma_file_info` might be a source of confusion, because this stuff only applies to `.xz` files. (Sidenote: For `.lzma` files, this gets a lot “easier”, because you can just peek at the bytes in the range `file[5:13]` to get an int64_t for that. Most implementations set it to `-1`, though). But as long as this is not being included in liblzma, no worries.

**Comment by addaleax** [comment]
If you want, I’m going to include this now? I think it does not *need* more tweaking besides taking care of the `message_error` calls, but that should be easy.

**Comment by jviotti** [comment]
@addaleax Yeah sure, go on! I have no idea how to handle asynchronicity to probably better for you to handle that.

**Comment by addaleax** [comment]
@jviotti Great – just to clarify, you are fine with licensing your code as public domain, so that it would be eligible for inclusion in liblzma at some later point? :)

**Comment by jviotti** [comment]
Sure, go ahead! :)

**Comment by addaleax** [comment]
Some good news: I’m getting somewhere with this! It’s getting late but I’d expect to have this one good to go tomorrow.

**Comment by jviotti** [comment]
Exciting, looking forward to try it out!
